### DISCUSSION TITLE
How reinitialize the components

### DISCUSSION THREAD

**Discussion by f4xy** [orchidjs/tom-select#211] [General]
I have a select2 component in my form and I have surrounded the select with a wire:ignore.

```
<div wire:ignore>
    <select id="selected_tasks" name="selected_tasks" multiple="multiple" class="selected_tasks form-select block w-full transition duration-150 ease-in-out sm:text-sm sm:leading-5">
        @foreach($tasks as $task)
            <option value="{{$task->id}}">
                {{$task->title}}
             </option>
         @endforeach
    </select>
</div>
```
The problem I have is if I submit my form and it fails validation the select2 component no longer lets you add or remove items.

The following solves the problem：
```
document.addEventListener("livewire:load", () => {
	Livewire.hook('message.processed', (message, component) => {
		$('.form-select').select2()
    
	}); });
```

How can I use the following code if tom-slecet is used?

```
document.addEventListener("livewire:load", () => {
	Livewire.hook('message.processed', (message, component) => {
		new TomSelect("#select-state", {});
	}); });
```

But it doesn't work。

<img width="908" alt="截屏2021-09-21 下午8 31 50" src="https://user-images.githubusercontent.com/88174794/134170800-cbe3152a-361c-47fd-889c-e455a3a38d19.png">

**Comment by oyejorge** [comment]
> <select id="selected_tasks" ...
> ...
> new TomSelect("#select-state",{});

Looks like it should be ```new TomSelect("#selected_tasks",{});```

**Reply by f4xy** [reply]
not working .
`tom-select.ts:99 Uncaught (in promise) Error: Tom Select already initialized on this element`

**Comment by oyejorge** [comment]
The sync method might be what you're looking for

https://tom-select.js.org/docs/api/#sync

**Reply by f4xy** [reply]
First of all, thank you very much for providing such a powerful plug-in. 
It is more suitable for me than select2. She can get the value directly in livewire. 
However, as you prompted, I added the following code. 

```
var peopleSelect = new TomSelect('.tom-select');
    document.addEventListener("livewire:load", () => {
        Livewire.hook('message.processed', (message, component) => {
            peopleSelect.sync();
        });
    });

```
When I dynamically add row, the new row cannot be rendered and displayed normally.
<img width="870" alt="截屏2021-09-23 下午11 50 54" src="https://user-images.githubusercontent.com/88174794/134542289-e2b54c13-74a0-4ba0-8041-8edd11b0e042.png">
<img width="913" alt="截屏2021-09-23 下午11 59 16" src="https://user-images.githubusercontent.com/88174794/134542397-416a9d95-385d-4c74-836e-e4250a00792c.png">

However, if I use the following code when using select2, I can successfully reload and display.
```
document.addEventListener("livewire:load", () => {
	Livewire.hook('message.processed', (message, component) => {
		$('.select-people').select2()
    
	}); });
```

**Comment by andreasmatuszewski** [comment]
Did you find a solution or are you still using select2?

**Comment by maulvieyazid** [comment]
I want to give the solution I found.
first, you can't use wire:model in tom select because when you change the value, Livewire will make changes continuously without stopping.
so the solution is to use **Livewire.emit** event.
you need to use **let** instead of **const**, because you will do the initialization again on tom select later.

```
      let pendamping = new TomSelect('#pendamping', {
          maxOptions: false,
          sortField: {},
      });
      
      pendamping.on('change', function(value){
          Livewire.emit('changePendamping', value);
      });
```

And then in the component's class i register a listener and declare a function of the event

```
    protected $listeners = ['changePendamping'];
    
    public function changePendamping($value)
        {
           // logic to update the value in DB
           
            $this->dispatchBrowserEvent('pendamping-updated', [
                'type'          => 'success',
                'message'       => 'Dosen Pendamping Berhasil Diubah.',
                'nikPendamping' => $value
            ]);
        }
```
inside the function i call the dispatchBrowserEvent method to display the success message to browser. And then i add event listener in my blade

```
    window.addEventListener('pendamping-updated', event => {
        pendamping.destroy();
    
        pendamping = new TomSelect('#pendamping', {
            maxOptions: false,
            sortField: {},
        });
        pendamping.setValue(event.detail.nikPendamping);
    
        Toastify({
            text: event.detail.message,
            duration: 3000,
            close: true,
            gravity: "bottom",
            position: "left",
            backgroundColor: event.detail.type == 'success' ? "#4fbe87" : "#dc3545",
        }).showToast();
    })
```
Notice that when livewire re-renders the component, it won't initialize the tom select.
so what we have to do is, call the destroy() method on tom select then initialize it again and don't forget to use setValue(), because when tom select is initialized again, the previously selected value will be lost.

**Comment by ampedweb** [comment]
Hi guys I thought I would post an adapted solution I found to this based on the following stack overflow article [here](https://stackoverflow.com/questions/73890677/tomselect-refresh-options-in-dependent-dropdown-after-repeated-ajax-load)

I execute the following code AFTER an ajax call has resolved and I have updated the underlying ```<select>``` element

```js

//This is a filter for clearOptions (see docs) to ensure that my initial placeholder option with a value of 0 isn't removed and stays in the same position in the option list (first)
const clearFilter = (option,value) => value == 0;

//Remove the currently selected option
this.tomSelectInstance.clear();
//Clear all of the options that have currently been rendered by tom select in the dropdown
this.tomSelectInstance.clearOptions(clearFilter);
//Syncronise the plugin with the updated element
this.tomSelectInstance.sync();
//Here I set my selected value back to 0, which is my "Blank" placeholder option
this.tomSelectInstance.setValue(0)
```

You can read more on these methods in the Tom Select API docs here: https://tom-select.js.org/docs/api/

Hopefully this is helpful to someone facing similar problems! :-)
